{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "authorship_tag": "ABX9TyMCR1di1tFNelM2jlOEElO5",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/PintoBI/mutagenic_primer_design/blob/main/generatemutagenicprimer.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!pip install Bio\n",
        "!git clone https://github.com/PintoBI/mutagenic_primer_design.git"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "jliOLhMdDown",
        "outputId": "baeb0d3d-55d8-4cc6-c708-ad1cffab5084"
      },
      "execution_count": 1,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Collecting Bio\n",
            "  Downloading bio-1.6.0-py3-none-any.whl (279 kB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m279.4/279.4 kB\u001b[0m \u001b[31m3.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hCollecting biopython>=1.80 (from Bio)\n",
            "  Downloading biopython-1.81-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.1 MB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.1/3.1 MB\u001b[0m \u001b[31m14.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hRequirement already satisfied: requests in /usr/local/lib/python3.10/dist-packages (from Bio) (2.31.0)\n",
            "Requirement already satisfied: tqdm in /usr/local/lib/python3.10/dist-packages (from Bio) (4.66.1)\n",
            "Collecting mygene (from Bio)\n",
            "  Downloading mygene-3.2.2-py2.py3-none-any.whl (5.4 kB)\n",
            "Requirement already satisfied: pandas in /usr/local/lib/python3.10/dist-packages (from Bio) (1.5.3)\n",
            "Requirement already satisfied: pooch in /usr/local/lib/python3.10/dist-packages (from Bio) (1.8.0)\n",
            "Collecting gprofiler-official (from Bio)\n",
            "  Downloading gprofiler_official-1.0.0-py3-none-any.whl (9.3 kB)\n",
            "Requirement already satisfied: numpy in /usr/local/lib/python3.10/dist-packages (from biopython>=1.80->Bio) (1.23.5)\n",
            "Collecting biothings-client>=0.2.6 (from mygene->Bio)\n",
            "  Downloading biothings_client-0.3.1-py2.py3-none-any.whl (29 kB)\n",
            "Requirement already satisfied: python-dateutil>=2.8.1 in /usr/local/lib/python3.10/dist-packages (from pandas->Bio) (2.8.2)\n",
            "Requirement already satisfied: pytz>=2020.1 in /usr/local/lib/python3.10/dist-packages (from pandas->Bio) (2023.3.post1)\n",
            "Requirement already satisfied: platformdirs>=2.5.0 in /usr/local/lib/python3.10/dist-packages (from pooch->Bio) (4.0.0)\n",
            "Requirement already satisfied: packaging>=20.0 in /usr/local/lib/python3.10/dist-packages (from pooch->Bio) (23.2)\n",
            "Requirement already satisfied: charset-normalizer<4,>=2 in /usr/local/lib/python3.10/dist-packages (from requests->Bio) (3.3.2)\n",
            "Requirement already satisfied: idna<4,>=2.5 in /usr/local/lib/python3.10/dist-packages (from requests->Bio) (3.4)\n",
            "Requirement already satisfied: urllib3<3,>=1.21.1 in /usr/local/lib/python3.10/dist-packages (from requests->Bio) (2.0.7)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.10/dist-packages (from requests->Bio) (2023.7.22)\n",
            "Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.10/dist-packages (from python-dateutil>=2.8.1->pandas->Bio) (1.16.0)\n",
            "Installing collected packages: biopython, gprofiler-official, biothings-client, mygene, Bio\n",
            "Successfully installed Bio-1.6.0 biopython-1.81 biothings-client-0.3.1 gprofiler-official-1.0.0 mygene-3.2.2\n",
            "Cloning into 'mutagenic_primer_design'...\n",
            "remote: Enumerating objects: 21, done.\u001b[K\n",
            "remote: Counting objects: 100% (21/21), done.\u001b[K\n",
            "remote: Compressing objects: 100% (20/20), done.\u001b[K\n",
            "remote: Total 21 (delta 3), reused 0 (delta 0), pack-reused 0\u001b[K\n",
            "Receiving objects: 100% (21/21), 9.34 KiB | 9.34 MiB/s, done.\n",
            "Resolving deltas: 100% (3/3), done.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 3,
      "metadata": {
        "id": "fitIApfwBXLg"
      },
      "outputs": [],
      "source": [
        "from Bio.SeqUtils import MeltingTemp as mt\n",
        "from Bio import Seq\n",
        "import csv\n",
        "from Bio import SeqIO\n",
        "import random\n",
        "\n",
        "def parse_fasta(file_path):\n",
        "    # Parse the FASTA file and extract the sequence\n",
        "    with open(file_path, 'r') as file:\n",
        "        records = SeqIO.parse(file, 'fasta')\n",
        "        for record in records:\n",
        "            return str(record.seq)\n",
        "\n",
        "\n",
        "def calculate_tm(primer):\n",
        "    \"\"\"\n",
        "    Calculates the melting temperature (Tm) of a mutagenic primer .\n",
        "\n",
        "        primer (str): The primer sequence with X in the mismatches.\n",
        "\n",
        "    Returns:\n",
        "        float: The melting temperature (Tm) in degrees Celsius.\n",
        "    \"\"\"\n",
        "    tm =mt.Tm_GC(primer,valueset=2)\n",
        "    return tm\n",
        "\n",
        "\n",
        "def translate_dna(dna_sequence):\n",
        "\n",
        "    # Translates a DNA sequence to a protein sequence.\n",
        "\n",
        "    # Args:\n",
        "    #     dna_sequence (str): The DNA sequence.\n",
        "\n",
        "    # Returns:\n",
        "    #     str: The translated protein sequence.\n",
        "\n",
        "    coding_dna = Seq.Seq(dna_sequence)\n",
        "    protein_sequence = coding_dna.translate()\n",
        "    return str(protein_sequence)\n",
        "\n",
        "\n",
        "def find_codon_position(dna_sequence, protein_position, firstresid):\n",
        "\n",
        "    # Finds the start position of the codon and retrieves the corresponding codon for a given protein position.\n",
        "    # Args:\n",
        "    #     dna_sequence (str): The DNA sequence.\n",
        "    #     protein_position (int): The position of the protein.\n",
        "    #     first_residue (int): The position of the first residue in the protein.\n",
        "    # Returns:\n",
        "    #     tuple: A tuple containing the start position of the codon and the codon itself.\n",
        "\n",
        "    codon_start = (protein_position - firstresid ) * 3\n",
        "    codon_end = codon_start + 3\n",
        "    codon = dna_sequence[codon_start:codon_end]\n",
        "    return codon_start, codon\n",
        "\n",
        "\n",
        "def import_codon_table(csv_file):\n",
        "\n",
        "    # Imports the codon table from a CSV file.\n",
        "\n",
        "    # Args:\n",
        "    #     csv_file (str): The path to the CSV file.\n",
        "\n",
        "    # Returns:\n",
        "    #     dict: The codon table as a dictionary.\n",
        "\n",
        "    genetic_code = {}\n",
        "    with open(csv_file, 'r', encoding='utf-8-sig') as file:\n",
        "        reader = csv.reader(file)\n",
        "        for row in reader:\n",
        "            amino_acid = row[0]\n",
        "            codons = row[1:]\n",
        "            genetic_code[amino_acid] = codons\n",
        "    return genetic_code\n",
        "\n",
        "\n",
        "def get_codons(codon_table, amino_acid, current_codon):\n",
        "    # Retrieves the codons for a given amino acid from the codon table.\n",
        "\n",
        "    # Args:\n",
        "    #     codon_table (dict): The codon table.\n",
        "    #     amino_acid (str): The amino acid.\n",
        "\n",
        "    # Returns:\n",
        "    #     list: The list of codons for the amino acid.\n",
        "    amino_acid = amino_acid.upper()  # Convert amino acid to uppercase\n",
        "    current_codon = current_codon.upper()  # Convert current_codon to uppercase\n",
        "\n",
        "    if amino_acid in codon_table:\n",
        "        codons = codon_table[amino_acid]\n",
        "        min_differences = float('inf')\n",
        "        selected_codon = None\n",
        "        for codon in codons:\n",
        "            if codon:  # Skip empty codons\n",
        "                differences = sum(c1 != c2 for c1, c2 in zip(codon, current_codon))\n",
        "                #print(differences,codon)\n",
        "                if differences < min_differences:\n",
        "                    min_differences = differences\n",
        "                    selected_codon = codon\n",
        "        return selected_codon\n",
        "    else:\n",
        "        return []\n",
        "\n",
        "def design_primer(dna_sequence, codon_start, mutant_codon, length):\n",
        "    start_index = codon_start - length\n",
        "    end_index = codon_start + 3 + length\n",
        "    primer_sequence = \"\".join([dna_sequence[start_index:codon_start], mutant_codon, dna_sequence[codon_start+3:end_index]])\n",
        "    primer_seq = Seq.Seq(primer_sequence)\n",
        "\n",
        "    return primer_seq\n",
        "\n",
        "\n",
        "def save_primers_to_csv(primer_dict):\n",
        "    # Save the primer dictionary to a CSV file\n",
        "    file_path = 'primers.csv'  # Specify the desired file path\n",
        "\n",
        "    with open(file_path, 'w', newline='') as file:\n",
        "        writer = csv.writer(file)\n",
        "        writer.writerow(['Primer Name', 'Sequence'])  # Write the header row\n",
        "\n",
        "        for primer_name, primer_seq in primer_dict.items():\n",
        "            writer.writerow([primer_name, primer_seq])  # Write each primer entry\n",
        "\n",
        "\n",
        "\n",
        "def parse_mutations(file_path):\n",
        "    # Parse the mutations file and extract the mutations\n",
        "\n",
        "    mutations = []\n",
        "    with open(file_path, 'r') as file:\n",
        "        reader = csv.DictReader(file)\n",
        "        for row in reader:\n",
        "            position = row['Position']\n",
        "            mutation = row['Mutation']\n",
        "            mutations.append((position, mutation))\n",
        "\n",
        "    return mutations\n",
        "\n",
        "def save_primers_to_csv(primer_dict, file_path):\n",
        "    # Save the primer dictionary to a CSV file\n",
        "    with open(file_path, 'w', newline='') as file:\n",
        "        writer = csv.writer(file)\n",
        "        writer.writerow(['Mutation', 'Primer Sequence', 'Tm'])  # Write the header row\n",
        "\n",
        "        for mutation, (primer_seq, tm) in primer_dict.items():\n",
        "            writer.writerow([mutation, str(primer_seq), tm])  # Write each primer entry\n",
        "\n",
        "def convert_mismatches(primer_seq, target_seq):\n",
        "    # Convert mismatched bases to 'X' in the primer sequence\n",
        "    converted_seq = ''\n",
        "    for p_base, t_base in zip(primer_seq, target_seq):\n",
        "        if p_base != t_base:\n",
        "            converted_seq += 'X'\n",
        "        else:\n",
        "            converted_seq += p_base\n",
        "    return converted_seq\n"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Example usage\n",
        "csv_file = '/content/mutagenic_primer_design/example/codontableXlaevis.csv'#high frequency codons for X_laevis\n",
        "codon_table = import_codon_table(csv_file)\n",
        "\n",
        "# Call the import function to start the process and save the cds_sequence\n",
        "cds_sequence = parse_fasta(\"/content/mutagenic_primer_design/example/Shaker_FL.fa\")# File path to a fasta sequence containing the coding DNA sequence\n",
        "protein_sequence = translate_dna(cds_sequence)\n",
        "starting_residue=1 #residue number of the first codon\n",
        "\n",
        "#select mutations\n",
        "mutations = parse_mutations(\"/content/mutagenic_primer_design/example/mutations.csv\") # File path to a CSV file containing the desired mutations\n",
        "threshold_tm = 78.0  # Set the threshold temperature here (78 works ok)\n",
        "max_length=26 # max primer length is 2*max_length+3\n",
        "\n",
        "# Create a dictionary to store the generated primers\n",
        "primer_dict = {}\n",
        "\n",
        "for mutation in mutations:\n",
        "    position, mutation = mutation\n",
        "    codonstart, mutatecodon = find_codon_position(cds_sequence, int(position), starting_residue)\n",
        "    mutant_codon = get_codons(codon_table, mutation, mutatecodon)\n",
        "    tm = 0.0  # Initialize tm to start the loop\n",
        "    length = 7  # Initial length value\n",
        "    while tm < threshold_tm and length <= max_length:\n",
        "        length += 1  # Increase the length by 1\n",
        "        primer_sequence = design_primer(cds_sequence, codonstart, mutant_codon, length).upper()\n",
        "        primer_target= Seq.Seq(cds_sequence[codonstart - length:codonstart+3+length]).upper()\n",
        "        mismatch_primer=convert_mismatches(primer_sequence,primer_target)\n",
        "        tm = calculate_tm(mismatch_primer)\n",
        "        primer_reverse_complement = str(Seq.Seq(primer_sequence).reverse_complement())\n",
        "    primer_dict['_'.join([position, mutation,'Fw'])] = primer_sequence, tm\n",
        "    primer_dict['_'.join([position, mutation,'Rv'])]=primer_reverse_complement, tm\n",
        "\n",
        "# Specify the desired file path to save the primers\n",
        "csv_file_path = 'primers.csv'\n",
        "\n",
        "# Call the function to save the primer dictionary as a CSV file\n",
        "save_primers_to_csv(primer_dict, csv_file_path)"
        "# Download results"
        "from google.colab import files"
        "files.download(csv_file_path)"
      ],
      "metadata": {
        "id": "OZwqG-WzD5AK"
      },
      "execution_count": 4,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "idKuculTEB65"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}
